Index: econet-trx-7.3.267.700/tools/trx/trx.h
===================================================================
--- econet-trx-7.3.267.700.orig/tools/trx/trx.h	2021-02-07 15:03:19.214512107 +0800
+++ econet-trx-7.3.267.700/tools/trx/trx.h	2021-02-07 15:03:19.222512107 +0800
@@ -13,8 +13,16 @@
 #endif
 
 #if 1//defined(ZYXEL)
+#if defined(TCSUPPORT_SECURE_BOOT_V2)
+#define ECONET_TRX_HEADER_LENGTH	0x900
+#define ZYXEL_TRX_HEADER_LENGTH		0x974	//sizeof(struct trx_header)
+#elif defined(TCSUPPORT_SECURE_BOOT_V1) || defined(TCSUPPORT_SECURE_BOOT_FLASH_OTP)
+#define ECONET_TRX_HEADER_LENGTH	0x214
+#define ZYXEL_TRX_HEADER_LENGTH		0x288	//sizeof(struct trx_header)
+#else
 #define ECONET_TRX_HEADER_LENGTH	0x100
 #define ZYXEL_TRX_HEADER_LENGTH		0x174	//sizeof(struct trx_header)
+#endif
 #define CHIP_ID_LEN 		8
 #define BOARD_ID_LEN 		16
 #define SW_VERSION_LEN 		32
Index: econet-trx-7.3.267.700/tools/trx/trx_config
===================================================================
--- econet-trx-7.3.267.700.orig/tools/trx/trx_config	2020-11-06 17:16:14.000000000 +0800
+++ econet-trx-7.3.267.700/tools/trx/trx_config	2021-02-08 03:57:50.424362787 +0800
@@ -1,3 +1,3 @@
-version=7.3.141.0_v001
+version=7.3.267.1
 customerversion=
 Model=3 6035 122 0
