Index: gdb-linaro-7.5-2012.12-1/sim/common/sim-arange.h
===================================================================
--- gdb-linaro-7.5-2012.12-1.orig/sim/common/sim-arange.h
+++ gdb-linaro-7.5-2012.12-1/sim/common/sim-arange.h
@@ -62,17 +62,13 @@ extern void sim_addr_range_delete (ADDR_
 
 /* Return non-zero if ADDR is in range AR, traversing the entire tree.
    If no range is specified, that is defined to mean "everything".  */
-extern INLINE int
+static INLINE int
 sim_addr_range_hit_p (ADDR_RANGE * /*ar*/, address_word /*addr*/);
 #define ADDR_RANGE_HIT_P(ar, addr) \
   ((ar)->range_tree == NULL || sim_addr_range_hit_p ((ar), (addr)))
 
 #ifdef HAVE_INLINE
-#ifdef SIM_ARANGE_C
-#define SIM_ARANGE_INLINE INLINE
-#else
-#define SIM_ARANGE_INLINE EXTERN_INLINE
-#endif
+#define SIM_ARANGE_INLINE static INLINE
 #include "sim-arange.c"
 #else
 #define SIM_ARANGE_INLINE
