Index: linux-4.1.52/net/core/skbuff.c
===================================================================
--- linux-4.1.52.orig/net/core/skbuff.c	2020-10-30 11:24:06.104818299 +0800
+++ linux-4.1.52/net/core/skbuff.c	2020-10-30 12:28:57.441915300 +0800
@@ -1755,6 +1755,16 @@
 	recycle_hook = dst->recycle_hook;
 	recycle_context = dst->recycle_context;
 
+	if (unlikely((src->recycle_flags & SKB_DATA_RECYCLE) &&
+	   ((recycle_hook != src->recycle_hook) ||
+	    (recycle_context != src->recycle_context))))
+	{
+	    /* free the skb->head from src and reallocate from kernel
+	     * if pskb_expand_head returns fail, unhandled error will be triggered.
+	     * so BUG_ON here. */
+	    BUG_ON(pskb_expand_head(src, 0, 0, GFP_ATOMIC));
+	}
+
 	skb = __skb_clone(dst, src);
 
 	dst->recycle_flags |= recycle_flags;
