Index: linux-4.1.52/Kconfig.zyxel
===================================================================
--- linux-4.1.52.orig/Kconfig.zyxel	2021-02-07 14:24:47.333892417 +0800
+++ linux-4.1.52/Kconfig.zyxel	2021-02-08 13:13:50.540501353 +0800
@@ -33,3 +33,6 @@
 
 config ZYXEL_PPTPD_RELAY
         bool "ZyXEL feature - PPTP relay for PPPOA"
+
+config ZYXEL_L2TP_SUPPORT
+        bool "Support L2TP VPN"
\ No newline at end of file
Index: linux-4.1.52/drivers/net/ppp/ppp_generic.c
===================================================================
--- linux-4.1.52.orig/drivers/net/ppp/ppp_generic.c	2021-02-07 14:24:47.321892417 +0800
+++ linux-4.1.52/drivers/net/ppp/ppp_generic.c	2021-02-08 13:15:27.151422351 +0800
@@ -3073,6 +3073,12 @@
          sprintf(dev->name, "pppowwan%d", num[0]);
       }
 #endif
+#if defined(CONFIG_ZYXEL_L2TP_SUPPORT)
+      else if (num[1] == 5) /* l2tp */
+      {
+        sprintf(dev->name, "l2tp%d", num[0]);
+      }
+#endif
    }
 #else
 	sprintf(dev->name, "ppp%d", unit);
