Index: linux-4.1.52/drivers/mtd/nand/bcm63xx_spinand.c
===================================================================
--- linux-4.1.52.orig/drivers/mtd/nand/bcm63xx_spinand.c	2021-03-19 18:06:55.187413029 +0800
+++ linux-4.1.52/drivers/mtd/nand/bcm63xx_spinand.c	2021-03-22 09:53:40.523334887 +0800
@@ -483,6 +483,26 @@
         .chip_subpage_shift = 2, // 2^ shift amount based on number of subpages (4)
         .chip_clock_speed = 104000000,
     },
+    { // 2Gb,
+        .chip_name = "Winbond W25N02KV",
+        .chip_device_id = {WINBONDPART, ID_W25N02KV_1, ID_W25N02KV_2, 0},
+        .chip_page_size = 2048,
+        .chip_page_shift = 11,
+        .chip_block_size = 64 * 2048, // 64 pages per block x chip_page_size
+        .chip_block_shift = 17,       // block size of 0x20000 (128KB)
+        .chip_spare_size = 128,
+        .chip_ecc_offset = 0x800,     // location of ECC bytes
+        .chip_num_blocks = 2048,      // 2048 blocks total
+        .chip_num_planes = 1,
+        .chip_die_sel = 0,    // no support die select
+        .chip_total_size = 64 * 2048 * 2048, // chip_block_size x chip_num_blocks
+        .ecclayout =&spinand_oob_toshiba_micron_ab,
+        .chip_ecc = 8, //  8-Bit ECC bits
+        .chip_ecc_corr = 6, // threshold to fix correctable bits (6/8)
+        .chip_ecc_enh = 0, // enhanced bad bit detection by chip (none)
+        .chip_subpage_shift = 2, // 2^ shift amount based on number of subpages (4)
+        .chip_clock_speed = 104000000,
+    },
     { // 1Gb
         .chip_name = "Macronix MX35LF1GE4",
         .chip_device_id = {MACRONIXPART, ID_MX35LF1GE4, 0, 0},
