Index: linux-4.1.52/arch/arm/include/asm/memory.h
===================================================================
--- linux-4.1.52.orig/arch/arm/include/asm/memory.h
+++ linux-4.1.52/arch/arm/include/asm/memory.h
@@ -39,7 +39,7 @@
  * TASK_SIZE - the maximum size of a user space task.
  * TASK_UNMAPPED_BASE - the lower boundary of the mmap VM area
  */
-#define TASK_SIZE		(UL(CONFIG_PAGE_OFFSET) - UL(SZ_16M))
+#define TASK_SIZE		(UL(CONFIG_PAGE_OFFSET) - UL(SZ_18M))
 #define TASK_UNMAPPED_BASE	ALIGN(TASK_SIZE / 3, SZ_16M)
 
 /*
@@ -52,7 +52,7 @@
  * and PAGE_OFFSET - it must be within 32MB of the kernel text.
  */
 #ifndef CONFIG_THUMB2_KERNEL
-#define MODULES_VADDR		(PAGE_OFFSET - SZ_16M)
+#define MODULES_VADDR		(PAGE_OFFSET - SZ_18M)
 #else
 /* smaller range for Thumb-2 symbols relocation (2^24)*/
 #define MODULES_VADDR		(PAGE_OFFSET - SZ_8M)
Index: linux-4.1.52/include/linux/sizes.h
===================================================================
--- linux-4.1.52.orig/include/linux/sizes.h
+++ linux-4.1.52/include/linux/sizes.h
@@ -35,6 +35,7 @@
 #define SZ_4M				0x00400000
 #define SZ_8M				0x00800000
 #define SZ_16M				0x01000000
+#define SZ_18M				0x01200000
 #define SZ_32M				0x02000000
 #define SZ_64M				0x04000000
 #define SZ_128M				0x08000000
