Index: ebtables-v2.0.8-2/Makefile
===================================================================
--- ebtables-v2.0.8-2.orig/Makefile	2022-01-19 14:10:37.958333000 +0800
+++ ebtables-v2.0.8-2/Makefile	2022-01-19 14:12:38.290333000 +0800
@@ -23,6 +23,8 @@
 SYSROOT:=/opt/trendchip/buildroot-gcc493_glibc222_arm32_32bServer/usr/arm-buildroot-linux-gnueabi/sysroot
 else ifeq ($(CONFIG_TC_SDK_VERSION),"7.3.267.750")
 SYSROOT:=/opt/trendchip/buildroot-gcc493_glibc222_arm32_32bServer/usr/arm-buildroot-linux-gnueabi/sysroot
+else ifeq ($(CONFIG_TC_SDK_VERSION),"7.3.267.700")
+SYSROOT:=/opt/trendchip/buildroot-gcc493_glibc222_arm32_32bServer/usr/arm-buildroot-linux-gnueabi/sysroot
 else
 SYSROOT:=/opt/trendchip/mipsel-linux-uclibc-4.6.3-kernel3.18/usr/mipsel-buildroot-linux-uclibc/sysroot
 endif
